src/sinks/appsignal/mod.rs

```diff
@@ -153,25 +203,40 @@ impl HttpSink for AppsignalSinkConfig {
     }
 
     async fn build_request(&self, events: Self::Output) -> crate::Result<Request<Bytes>> {
-        let uri = endpoint_uri(self.endpoint.clone(), "vector/events")?;
-        let request = Request::post(uri)
-            .header(
-                AUTHORIZATION,
-                format!("Bearer {}", self.push_api_key.inner()),
-            )
-            .header("Content-Encoding", "gzip");
-
-        let mut body = crate::serde::json::to_bytes(&events).unwrap().freeze();
-
-        let buffer = BytesMut::new();
-        let mut writer = GzEncoder::new(buffer.writer(), flate2::Compression::new(6));
-        writer.write_all(&body).expect("Writing to Vec can't fail");
-        body = writer
-            .finish()
-            .expect("Writing to Vec can't fail")
-            .into_inner()
-            .into();
+        let uri = endpoint_uri(&self.endpoint, "vector/events")?;
+        let mut request = Request::post(uri).header(
+            AUTHORIZATION,
+            format!("Bearer {}", self.push_api_key.inner()),
+        );
 
+        let mut body = crate::serde::json::to_bytes(&events)?.freeze();
+
+        match self.compression {
+            AppsignalCompression::Gzip => {
+                request = request.header("Content-Encoding", "gzip");
+
+                let buffer = BytesMut::new();
+                let mut writer = GzEncoder::new(buffer.writer(), flate2::Compression::new(6));
+                write_all(&mut writer, 0, &body)?;
+                body = writer
+                    .finish()
+                    .context(CompressionFailedSnafu)?
+                    .into_inner()
+                    .into();
+            }
+            AppsignalCompression::Zlib => {
+                request = request.header("Content-Encoding", "deflate");
+
+                let buffer = BytesMut::new();
+                let mut writer = ZlibEncoder::new(buffer.writer(), flate2::Compression::new(6));
+                write_all(&mut writer, 0, &body)?;
+                body = writer
+                    .finish()
+                    .context(CompressionFailedSnafu)?
+                    .into_inner()
+                    .into();
+            }
+        }
```

Comment:
A small code reduction opportunity- it looks like the only things unique about the two cases are the writer instantiation and the adding of the request header. You could collapse this into something like the following:


```suggestion
        let buffer = BytesMut::new();
        
        let mut writer = match self.compression {
            AppsignalCompression::Gzip => {
                request = request.header("Content-Encoding", "gzip");
                GzEncoder::new(buffer.writer(), flate2::Compression::new(6))
            }
            AppsignalCompression::Zlib => {
                request = request.header("Content-Encoding", "deflate");
                ZlibEncoder::new(buffer.writer(), flate2::Compression::new(6))
            }
        };
        
        write_all(&mut writer, 0, &body)?;
        body = writer
             .finish()
             .context(CompressionFailedSnafu)?
             .into_inner()
             .into();
```

Reply:
I couldn't get your exact example to work, even with type annotations. See the compiler output:

<details>
  <summary>Click for compiler output</summary>

  ```
  error[E0308]: `match` arms have incompatible types
    --> src/sinks/appsignal/mod.rs:222:17
      |
  215 |           let mut writer = match self.compression {
      |  __________________________-
  216 | |             AppsignalCompression::Gzip => {
  217 | |                 request = request.header("Content-Encoding", "gzip");
  218 | |                 GzEncoder::new(buffer.writer(), flate2::Compression::new(6))
      | |                 ------------------------------------------------------------ this is found to be of type `flate2::write::GzEncoder<bytes::buf::Writer<bytes::BytesMut>>`
  ...   |
  222 | |                 ZlibEncoder::new(buffer.writer(), flate2::Compression::new(6))
      | |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `flate2::write::GzEncoder`, found struct `flate2::write::ZlibEncoder`
  223 | |             }
  224 | |         };
      | |_________- `match` arms have incompatible types
      |
      = note: expected struct `flate2::write::GzEncoder<bytes::buf::Writer<bytes::BytesMut>>`
                found struct `flate2::write::ZlibEncoder<bytes::buf::Writer<bytes::BytesMut>>`

  error[E0599]: no method named `finish` found for trait object `dyn std::io::Write` in the current scope
    --> src/sinks/appsignal/mod.rs:228:14
      |
  228 |             .finish()
      |              ^^^^^^ method not found in `dyn std::io::Write`
      |
      = help: items from traits can only be used if the trait is implemented and in scope
      = note: the following traits define an item `finish`, perhaps you need to implement one of them:
              candidate #1: `sinks::util::batch::Batch`
              candidate #2: `sinks::prometheus::collector::MetricCollector`
  ```
</details>

I ended up using the Compression and Compressor struct to accomplish the same thing. Could be better because it now uses the Vector utils to compress the payload rather than directly interface with the flate2 crate.

Reply:
Ah, serves me right giving a suggestion without trying it :)

I will take a look at that but, I had a discussion with the @vectordotdev/ux-team  about this, and it seems that we generally want to give users the flexibility to decide whether or not to use compression, with the default being no compression. In which case, it seems like for this PR it would make sense to use the sink util `Compression` struct directly in the AppSignal config struct, and that would change this parsing significantly.

Reply:
> with the default being no compression

Log streams can get quite huge, and we've found that they compress really well. So I think having compression turned on by default is beneficial for everybody. Can you explain the reasoning behind no compression being the prefered default?

Reply:
> Can you explain the reasoning behind no compression being the prefered default?

Sure, here are some additional insights shared from the UX team. Compression is a trade off between throughput, resource utilization and network traffic. Our primary target has been to reduce throughput. To that goal we ran experiments with the Datadog components and making uncompressed the default resulted in the lowest throughput. That has thus become the default we aim to keep for consistency across sinks and in line with the goal of reducing throughput.

There doesn't seem to be anything unique about this new integration to warrant changing that for the introduction of this sink (unless we're missing something. However we would be open to considering re-evaluating the compression default at a holistic level but would opt for doing that as a separate effort from this PR. It is possible we could come up with a new consensus for default that would be a healthy compromise between performance and network traffic. This has briefly come up in the past regarding the Datadog components, but a new conclusion was not found.

Reply:
I've updated the PR to use the built-in Compression struct. It has no specified default, so I think that means no compression is used. I'm a bit confused though as other sinks do use gzip by default. Can we specify it as such on our sink as well? Or should be the other be updated to also use no compression?

Sinks with default gzip compression:

* [AWS S3](https://github.com/vectordotdev/vector/blob/56208cc9bded7fd5a12c827f5198f6c03e7f2a32/src/sinks/aws_s3/config.rs#L113-L114)
* [Azure blob](https://github.com/vectordotdev/vector/blob/56208cc9bded7fd5a12c827f5198f6c03e7f2a32/src/sinks/azure_blob/config.rs#L122-L123)
* [Clickhouse](https://github.com/vectordotdev/vector/blob/56208cc9bded7fd5a12c827f5198f6c03e7f2a32/src/sinks/clickhouse/config.rs#L46-L47)
* [Datadog traces](https://github.com/vectordotdev/vector/blob/56208cc9bded7fd5a12c827f5198f6c03e7f2a32/src/sinks/datadog/traces/config.rs#L155)
* [NewRelic](https://github.com/vectordotdev/vector/blob/56208cc9bded7fd5a12c827f5198f6c03e7f2a32/src/sinks/new_relic/config.rs#L99-L100)
* [Webhdfs](https://github.com/vectordotdev/vector/blob/56208cc9bded7fd5a12c827f5198f6c03e7f2a32/src/sinks/webhdfs/config.rs#L94-L95)

Reply:
That's a great observation... The "formal" UX team is still pretty new and thus we're sort of in a flux state where we are trying to establish better consistency across the components, and in this case we are indeed more inconsistent than we'd like. It seems this is a case where we will be discussing (independently of this PR) a holistic approach and documenting that.

In the meantime, since we do have a number of sinks using gzip as the default, we are fine with that being the default here, with the understanding that if we do an evaluation later on that it may be considered along with the rest of the sinks for a different default.

The most straightforward way to achieve that would probably just be to set a `#[serde(default =` like you have for `endpoint` , that specifies the gzip enum variant.

Sorry for the churn on that!

Reply:
Great! I've changed it to use gzip by default for now. I understand the situation that things may not be consistent everywhere yet and this may change in the future 👍